File: problems/2563-count-the-number-of-fair-pairs/solution.py
# Link: https://leetcode.cn/problems/count-the-number-of-fair-pairs/
# Level: Medium
# Date: 2025-04-19

class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        nums.sort()
        res = 0
        for j, x in enumerate(nums):
            r = bisect_right(nums, upper - x, 0, j)
            l = bisect_left(nums, lower - x, 0, j)
            res += r - l
        return res

File: problems/2563-count-the-number-of-fair-pairs/solution.cpp
// Link: https://leetcode.cn/problems/count-the-number-of-fair-pairs/
// Level: Medium
// Date: 2025-04-19

class Solution {
    public:
        long long countFairPairs(vector<int>& nums, int lower, int upper) {
            long long res = 0;
            sort(nums.begin(), nums.end());
            for (int j = 0; j < nums.size(); j++) {
                auto r = upper_bound(nums.begin(), nums.begin() + j, upper - nums[j]);
                auto l = lower_bound(nums.begin(), nums.begin() + j, lower - nums[j]);
                res += r - l;
            }
            return res;
        }
};
